src/common/dao/repository.go

```diff
@@ -124,46 +103,79 @@ func GetTopRepos(projectIDs []int64, n int) ([]*models.RepoRecord, error) {
-	qs := GetOrmer().QueryTable(&models.RepoRecord{})
-	if len(name) != 0 {
-		qs = qs.Filter("Name__contains", name)
+func GetTotalOfRepositories(query ...*models.RepositoryQuery) (int64, error) {
+	sql, params := repositoryQueryConditions(query...)
+	sql = `select count(*) ` + sql
+	var total int64
+	if err := GetOrmer().Raw(sql, params).QueryRow(&total); err != nil {
+		return 0, err
 	}
-	return qs.Count()
+	return total, nil
 }
 
-// GetTotalOfRepositoriesByProject ...
-func GetTotalOfRepositoriesByProject(projectIDs []int64, name string) (int64, error) {
-	if len(projectIDs) == 0 {
-		return 0, nil
-	}
-
-	qs := GetOrmer().QueryTable(&models.RepoRecord{}).
-		Filter("project_id__in", projectIDs)
+// GetRepositories ...
+func GetRepositories(query ...*models.RepositoryQuery) ([]*models.RepoRecord, error) {
+	repositories := []*models.RepoRecord{}
 
-	if len(name) != 0 {
-		qs = qs.Filter("Name__contains", name)
+	sql, params := repositoryQueryConditions(query...)
+	sql = `select r.repository_id, r.name, r.project_id, r.description, r.pull_count, 
+	r.star_count, r.creation_time, r.update_time ` + sql + `order by r.name `
+	if len(query) > 0 && query[0] != nil {
+		page, size := query[0].Page, query[0].Size
+		if size > 0 {
+			sql += `limit ? `
+			params = append(params, size)
+			if page > 0 {
+				sql += `offset ? `
+				params = append(params, size*(page-1))
+			}
+		}
 	}
 
-	return qs.Count()
+	if _, err := GetOrmer().Raw(sql, params).QueryRows(&repositories); err != nil {
+		return nil, err
+	}
+	return repositories, nil
 }
 
-// GetRepositoriesByProject ...
-func GetRepositoriesByProject(projectID int64, name string,
-	limit, offset int64) ([]*models.RepoRecord, error) {
+func repositoryQueryConditions(query ...*models.RepositoryQuery) (string, []interface{}) {
+	params := []interface{}{}
+	sql := `from repository r `
+	if len(query) == 0 || query[0] == nil {
+		return sql, params
+	}
+	q := query[0]
+	if len(q.ProjectName) > 0 {
+		sql += `join project p on r.project_id = p.project_id `
+	}
 
-	repositories := []*models.RepoRecord{}
+	if len(q.LabelIDs) > 0 {
+		sql += `join harbor_resource_label rl on r.repository_id = rl.resource_id 
+		and rl.resource_type = 'r' `
+	}
+	sql += `where 1=1 `
 
-	qs := GetOrmer().QueryTable(&models.RepoRecord{}).
-		Filter("ProjectID", projectID)
+	if len(q.Name) > 0 {
+		sql += `and r.name = ? `
+		params = append(params, q.Name)
+	}
 
-	if len(name) != 0 {
-		qs = qs.Filter("Name__contains", name)
+	if len(q.ProjectIDs) > 0 {
+		sql += fmt.Sprintf(`and r.project_id in ( %s ) `,
+			paramPlaceholder(len(q.ProjectIDs)))
+		params = append(params, q.ProjectIDs)
 	}
-	if limit > 0 {
-		qs = qs.Limit(limit).Offset(offset)
+
+	if len(q.ProjectName) > 0 {
+		sql += `and p.name = ? `
+		params = append(params, q.ProjectName)
 	}
-	_, err := qs.All(&repositories)
 
-	return repositories, err
+	if len(q.LabelIDs) > 0 {
+		sql += fmt.Sprintf(`and rl.label_id in ( %s ) `,
```

Comment:
does it mean the repo will be returned if it has one of the labels in the query parm?